### ISSUE TITLE
False positive error report for interface type missing index signature, thus not assignable to object type with `unknown` values.

### ISSUE THREAD

**Issue by woutervh-** [microsoft/TypeScript#32263]
**TypeScript Version:**  3.6.0-dev.20190704

**Search Terms:**
index signature, unknown, assignable

**Code**

```ts
interface A {
  properties?: { [Key: string]: unknown } | null;
}

type Bar = {
  key?: "value";
};

interface Baz {
  key?: "value";
}

interface B extends A { // ok
  properties?: Bar;
}

interface C extends A { // error: Baz is not assignable to { [Key: string]: unknown } -> Index signature is missing
  properties?: Baz;
}
```

**Expected behavior:**
No error expected, interface C extends A correctly.

**Actual behavior:**
Error because interface C does not extend interface A

**Playground Link:** 
http://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgILIN4ChnIA5QD2e0YwEAzgPwBcmyA2gNIQCedFYUoA5gLp0AriADWIQgHcQyAL7IAPshCCANioDcWGVixhWJZACE4UZAF5MOZCLa1kAIgBucFYIj3NMzVlCRYiFGMAL0tcG1Y7Jxc3Dy0dX2h4JCNkCAAPSBAAEwo0egB6fORCESsCYlJyajpjKE948ESA5ABhVIyIbNz0DGRC1KgiKBq4EOBc8TBkOAoKYB4QOAAjFRQwQnpmNg4uXgFkYTFJaTkAWgA+ZABJbPTkOYW4MEEoFHHkAFtxuZAeMqISFAyJQ7MF6kA

**Related Issues:** 
#30977

**Comment by jack-williams** [comment]
I believe this is working as intended: object literal types have implicit index signatures, while interfaces do not. See the implementation here: #7029

**Comment by woutervh-** [comment]
Thanks @jack-williams , didn't know about this specific assignability rule. Do you know if there is a comprehensive list of type assignability rules out there? I tried to search for one but couldn't find.

**Comment by jack-williams** [comment]
Not really - If you want to know the absolute truth you need to go here: 

- https://github.com/microsoft/TypeScript/blob/master/src/compiler/checker.ts

The spec is pretty good, although incomplete:

- https://github.com/microsoft/TypeScript/blob/master/doc/spec.md#3114-assignment-compatibility

And something to keep an eye on:

- https://github.com/microsoft/TypeScript-New-Handbook

**Comment by fatcerberus** [comment]
I hate that `checker.ts` is so absolutely ginormous that GitHub refuses to display it - it makes it impossible to link to specific lines.

**Comment by jack-williams** [comment]
The reason intellisense for TypeScript is so performant is because the TypeScript team want to have auto-completion while editing `checker.ts`[1] - so if anything, we should be encouraging them to make it larger.
 


<sub> [1] I made that up</sub>

**Comment by RyanCavanaugh** [comment]
I need to add this to the interfaces vs type aliases section.

> ...  if anything, we should be encouraging them to make it larger.

I feel seen. Also we should make the binder incremental.
